### PR TITLE
fix(provider/kubernetes): fix k8s client configuration and image id p…

### PR DESCRIPTION
**Pull request by chlung**
About this PR,

Fix the problem when context does not present in k8s configuration file and incorrect image id assignment.

…roblem.

We prefer small, well tested pull requests.

Please refer to [Contributing to Spinnaker](https://spinnaker.io/community/contributing/).

When filling out a pull request, please consider the following:

* Follow the commit message conventions [found here](http://www.spinnaker.io/v1.0/docs/how-to-submit-a-patch).
* Provide a descriptive summary for your changes.
* If it fixes a bug or resolves a feature request, be sure to link to that issue.
* Add inline code comments to changes that might not be obvious.
* Squash your commits as you keep adding changes.
* Add a comment to @spinnaker/reviewers for review if your issue has been outstanding for more than 3 days.

Note that we are unlikely to accept pull requests that add features without prior discussion. The best way to propose a feature is to open an issue first and discuss your ideas there before implementing them.

### CODE REVIEW

**Comment by chlung** [comment]
@lwander Please review.  Thanks.

**Comment by ethanfrogers** [comment]
@lwander @chlung I wonder if this is really an issue with Spinnaker. It seems like the client library is just assuming that `current-context` exists. If the client libraries across various languages require it, maybe we should say that the provided Kubernetes config is invalid. If it isn't, which it wasn't with the Fabric8 library, the client library should be fixed accordingly.

**Comment by chlung** [comment]
@lwander  and @ethanfrogers.  Without these required fields, the kubectl won't work.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy [hunk 1]
```diff
@@ -22,18 +22,48 @@ import io.kubernetes.client.util.KubeConfig;
...
+      configMap.put("clusters", cluster)
+    }
+
+    if (this.user) {
+      configMap.put("users", user)
+    }
+
+    yaml = new Yaml(new SafeConstructor());
+    String out = yaml.dump(configMap);
+    InputStream kubeconfigIS = new ByteArrayInputStream(out.getBytes());
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy]: What is the goal here? You don't need to recreate a kubeconfig. If a user has specified a context, it's assumed it exists in their kubeconfig, not the one they are creating.

Comment (chlung) [reply clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy]: If the value is presenting is clouddriver-local.yml.  It should be allow to override.  Is it? or we just need to update if the value does not present in the config file.

**Comment by lwander** [comment]
That's a good point @chlung, also this is a regression, so even if it wasn't needed it's still breaking users (like yourself) without warning.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy (hunk 1 above)
Comment (lwander) [reply clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy]: Yes it should override `current-context`. That doesn't mean we need to create a new kubeconfig object, though.

**Comment by chlung** [comment]
I can throw some error message if these fields are missing in both config and clouddriver-local.yml.

**Comment by chlung** [comment]
@lwander https://github.com/kubernetes-client/java/blob/master/util/src/main/java/io/kubernetes/client/util/KubeConfig.java#L101 will still throw NPE if "current-context" does not present.  So, we must re-create the kubeconfig with all required data before call or allow call setContext() later if null.  Please review and let me your suggestion.

**Comment by lwander** [comment]
I don't see how this will work - if the kubeconfig you create doesn't have any credentials beyond the name of the context, cluster & user you add how will it authenticate?

**Comment by chlung** [comment]
Then, we might consider to call setContext() later if the initial context is null.

**Comment by lwander** [comment]
I would just follow the pattern in the code I sent you via slack

Here is the relevant section: https://github.com/spinnaker/clouddriver/blob/master/clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/v2/security/KubernetesV2Credentials.java#L127-L146

**Comment by chlung** [comment]
@lwander I patched the changes per our discussion.  Please merge this PR after client library patched and apply to clouddriver-kubernetes; otherwise, the reported issue won't fix.  thanks.

File: clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy [hunk 2]
```diff
@@ -22,18 +22,65 @@ import io.kubernetes.client.util.KubeConfig;
...
+
+    try {
+      if (StringUtils.isEmpty(kubeconfigFile)) {
+        kubeconfig = KubeConfig.loadDefaultKubeConfig()
+      } else {
+        kubeconfig = KubeConfig.loadKubeConfig(new FileReader(kubeconfigFile))
+      }
+    } catch (FileNotFoundException e) {
+      throw new RuntimeException("Unable to create credentials from kubeconfig file: " + e, e)
+    } catch (Exception e2) {
```
Comment (lwander) [inline clouddriver-kubernetes/src/main/groovy/com/netflix/spinnaker/clouddriver/kubernetes/security/KubernetesApiClientConfig.groovy]: You should probably handle this as well ( you can just throw an exception )

**Comment by ethanfrogers** [comment]
Out of curiosity, what's the patch that needs to be made to the client library?

**Comment by lwander** [comment]
Client library has the same issue where if `current-context` isn't set, it will crash. We could get around it here by loading the kubeconfig by hand & setting it if unset, but that would inconsistent with how kubectl handles it (by creating the default localhost:8080 context).

**Comment by chlung** [comment]
@lwander I updated the code per your feedback.  Please review.  thanks.

**Comment by chlung** [comment]
@ethanfrogers thanks

**Pull request merged**